lib/crypto/aead.lua

```diff
@@ -0,0 +1,132 @@
+--
+-- SPDX-FileCopyrightText: (c) 2025 jperon <cataclop@hotmail.com>
+-- SPDX-License-Identifier: MIT OR GPL-2.0-only
+--
+
+--- AEAD (Authenticated Encryption with Associated Data) operations.
+-- This module provides a Lua wrapper for AEAD operations,
+-- using the underlying 'crypto_aead' C module.
+--
+-- @module crypto.aead
+
+local new = require("crypto_aead").new
+local sub = string.sub
+
+--- AEAD instance methods.
+-- This table represents an AEAD instance configured for a specific algorithm.
+-- It is created by calling the main function of the `crypto.aead` module.
+-- @type AeadCipherInstance
+-- @field setkey Sets the encryption key.
+-- @field setauthsize Sets the authentication tag size.
+-- @field ivsize Gets the required nonce/IV size.
+-- @field authsize Gets the authentication tag size.
+-- @field encrypt Encrypts plaintext.
+-- @field decrypt Decrypts ciphertext.
+-- @field close Release resources associated with this AEAD instance.
+
+--- Creates a new AEAD cipher instance.
+-- This function acts as the constructor for AEAD instances.
+-- @tparam string algname The algorithm name, e.g., "gcm(aes)".
+-- @treturn AeadCipherInstance An AEAD instance table with methods for encryption and decryption.
+-- @raise Error if `algname` is not a string or if C object creation fails.
+-- @usage local gcm_aes = require("crypto.aead")("gcm(aes)")
+return function (algname)
+	if type(algname) ~= "string" then error("crypto.aead: constructor: algname must be a string", 2) end
+
+	local tfm = new(algname) -- This is the C userdata object
+
+	--- Sets the encryption key.
+	-- @function setkey
+	-- @within AeadCipherInstance
+	-- @tparam string key The encryption key.
+	-- @raise Error if key is not a string or if C operation fails.
+	local function setkey(key)
+		if type(key) ~= "string" then error("crypto.aead: setkey: key must be a string", 2) end
```

Comment:
it's not necessary for checking for types in the lua layer as you are already checking it in the C binding, right?